Doc: Pacemaker Administration: Document client environment variables

nrwahl2 opened the pull request:
A few thoughts:
* Some of these have no valid use case for an external client but are included for completeness ("If these are set for any reason, Pacemaker will check them and care about them under some circumstances.")
* A few are regular shell variables that Pacemaker checks. We could optionally add a note saying that they're standard shell variables.
* We could move the "Advanced use only" ones (the ones with *really* no valid use cases) to the end of the table instead of intermingling them alphabetically.

kgaillot commented on doc/sphinx/Pacemaker_Administration/options.rst:
```diff
@@ -0,0 +1,233 @@
+Additionally, Pacemaker sets the ``PCMK_service`` environment variable to the
+subsystem or tool name. External scripts such as resource agents and fence
+agents can check this variable to see whether they're being run by (for example)
+the executor or the fencer.
```
Let's move PCMK_service, CRM_meta_container_attribute_target, CRM_meta_physical_host, and OCF_RESKEY_CRM_meta_on_node into a separate new table of variables set by Pacemaker for OCF resource agent use, in the OCF resource agents section. We can add a blanket warning that users should not set any of these, and if they do, behavior is undefined; that Pacemaker may define other variables, but their names, values, and meanings are subject to change; and that non-OCF agents should not rely on any of these being set.

We can drop the `OCF_RESKEY_ip` example from Pacemaker Explained, and instead add an entry for `OCF_RESKEY_*` to the new table. We can also move all other references to `OCF_RESKEY` (clone variables etc.) from Pacemaker Explained.

For PCMK_service, let's say that the values are subject to change, and so useful mainly for logging (i.e. don't say they should check the variable).

That will cleanly separate cluster configuration and agent development, and give agent developers everything they need in one place.

kgaillot replied:
Bah, I didn't know that. For their purposes, `!= ""` would be sufficient, and better than `== "pacemaker-fenced"`.

nrwahl2 replied:
That depends. Anything that calls `pcmk__cli_init_logging()`, `crm_log_init()`, or `crm_log_preinit()` with a non-NULL first argument gets `PCMK_service` set (to `crm_system_name`). That includes most or all Pacemaker CLI tools but seemingly *not* the CTS helpers. (I haven't run them to make sure `PCMK_service` is unset.)

If checking whether the fence agent is being called by *any* Pacemaker tool (besides `cts-fence-helper`) is sufficient, then checking for non-empty is sufficient. If the fencing library wants to know whether it's called by a daemon, I'm not sure if there's something else it can check.

nrwahl2 replied:
> We can also move all other references to `OCF_RESKEY` (clone variables etc.) from Pacemaker Explained.

There are 116 of them. We could move the entire section 11.2.9 ("Clone Resource Agent Requirements") from Pacemaker Explained to Pacemaker Administration.

The division of resource documentation into Administration and Explained feels more arbitrary than usual, and I struggle to draw lines between them. I guess, very loosely, info that's more relevant to **writing resource agents** (e.g., what environment variables are exposed) can go into Administration, while info pertaining to generic resource **behavior** and **configuration** can go into Explained.

nrwahl2 replied:
> We could move the entire section 11.2.9 ("Clone Resource Agent Requirements") from Pacemaker Explained to Pacemaker Administration.

I think that's the way to go.

nrwahl2 commented on doc/sphinx/Pacemaker_Administration/options.rst:
```diff
@@ -0,0 +1,233 @@
+       :ref:`$CIB_shadow_dir <CIB_shadow_dir>`. Behaves similarly to
+       :ref:`CIB_file <CIB_file>` but doesn't require specifying a complete
+       path. Does not create the shadow file automatically. Typically set and
+       used by :ref:`crm_shadow <crm_shadow>`.
```
Sure. I didn't know that was the case. It definitely can be set explicitly after calling `crm_shadow --batch` -- in that case, the `crm_shadow` output instructs the user to do so.

kgaillot replied:
Ah true, I guess we can leave that out then. I was just trying to avoid boxing us into the current implementation forever.

nrwahl2 replied:
We could still say something about only setting it to the name of a shadow instance created by `crm_shadow`.

nrwahl2 commented:
Edited description of `CIB_shadow` variable per review.

TODO:
* Move discussion of obscure meta variables to the agents section of Pacemaker Administration (in the current draft, they're in the client variable table).
* Add a warning near the client variables table that the user should not set `OCF_RESKEY_*` (or maybe `OCF_*`) variables or `CRM_meta_*` variables manually, that Pacemaker sets them, that behavior when setting them manually is undefined, etc.
* Possibly add another table for variables that Pacemaker sets, to include `OCF_RESKEY_*`, etc. I'm not sure yet how far we want to go in that direction... the impetus was to note three specific variables that Pacemaker both sets and checks (plus `PCMK_service`, which it only sets). I don't want to get sidetracked into fleshing out a "how to write a resource agent" guide.

kgaillot commented:
> Edited description of `CIB_shadow` variable per review and added a few commits on top.

The LSB index block got split from its section
 
> TODO:
> 
>     * Add a warning near the client variables table that the user should not set `OCF_RESKEY_*` (or maybe `OCF_*`) variables or `CRM_meta_*` variables manually, that Pacemaker sets them, that behavior when setting them manually is undefined, etc.
> 
>     * Move discussion of obscure meta variables to the agents section of Pacemaker Administration (in the current draft, they're in the client variable table).

Yes, I'm thinking the OCF agents subsection specifically. Only OCF agents use environment variables for parameters (fence agents are the only other type that even has parameters, and they use stdin), and only OCF agents support unique and promotable clones. LSB and fence agents could take advantage of environment variables that happen to always be set, but I don't think we should advertise that as a supported API.
       
>       * Possibly add another table for variables that Pacemaker sets, to include `OCF_RESKEY_*`, etc. I'm not sure yet how far we want to go in that direction... the impetus was to note three specific variables that Pacemaker both sets and checks (plus `PCMK_service`, which it only sets). I don't want to get sidetracked into fleshing out a "how to write a resource agent" guide.

Yep I was thinking of moving all of those from the general client variables table to a new table in the OCF agents subsection. For simplicity the clone notification variables can be left in the form they are now.

>       * Could make a reasonable effort based on `add_action_env_vars()`. I'll have to do some tracing to figure out how some `CRM_meta_*` vars (namely `XML_RSC_ATTR_TARGET` and `PCMK__ENV_PHYSICAL_HOST`, see `pcmk__node_attr_target()`) get into the environment without the `OCF_RESKEY` prefix while others get the prefix.

The ones in that function and the others that we already know about could be worthwhile.

nrwahl2 commented:
Updated to address review:
* Fix splitting of LSB index
* Add OCF environment variable table

Added a commit on top to note that the default log file differs for Pacemaker Remote nodes.

---

Concerns:
* We're covering all environment variables set for OCF agents but not for alert scripts. Many are largely similar but without prefixes in the alert case. Some of the alert-specific ones are described separately in the alerts section.
* `PCMK_service`: Useful outside the context of OCF resource agents where we're documenting it here. For example, the fencing library currently checks it (and uses a specific value), as we've discussed.
* Some of these (at least `OCF_EXIT_REASON_PREFIX`, `OCF_TRACE_FILE`, `OCF_TRACE_RA`) are not part of the OCF spec. Opened T739.
* There are other environment variables that Pacemaker sets and a client probably should not set; but these are **not** set for use by OCF resource agents. Not specifically documented anywhere.
    * `CRM_notify_*` (in `crm_mon.c:send_custom_trap()`)
    * `LC_ALL`
    * `PCMK_cluster_type`
    * `PCMK_logfile` (in `remote_pidone.c`)
    * `PCMK_node_start_state`
    * `PCMK_remote_port`
    * `PCMK_respawned`

nrwahl2 commented:
Added commit on top for `CIB_pam_service`, `SBD_SYNC_RESOURCE_STARTUP`, and `SBD_WATCHDOG_TIMEOUT` in Pacemaker Explained.

kgaillot reviewed (commented):
Looks great

> Concerns:
> 
>     * We're covering all environment variables set for OCF agents but not for alert scripts. Many are largely similar but without prefixes in the alert case. Some of the alert-specific ones are described separately in the alerts section.

I think that's fine for now

> 
>     * `PCMK_service`: Useful outside the context of OCF resource agents where we're documenting it here. For example, the fencing library currently checks it (and uses a specific value), as we've discussed.

You could do a separate section for environment variables set for all agent types if you want, but I don't think it's terribly important right now

> 
>     * Some of these (at least `OCF_EXIT_REASON_PREFIX`, `OCF_TRACE_FILE`, `OCF_TRACE_RA`) are not part of the OCF spec. Opened T739.
> 
>     * There are other environment variables that Pacemaker sets and a client probably should not set; but these are **not** set for use by OCF resource agents. None of these are specifically documented anywhere.
>       
>       * `CRM_notify_*` (in `crm_mon.c:send_custom_trap()`)
>       * `LC_ALL`

This is just set for Pacemaker's internal use since it affects the behavior of the standard C library. A bunch of variables will be set simply because Pacemaker doesn't remove variables after forking a child, but that shouldn't be considered providing them to the agent. Basically we just want to document what we're effectively guaranteeing as a public API.

>       * `PCMK_cluster_type`
>       * `PCMK_logfile` (in `remote_pidone.c`) (not documented for setting, but this one is borderline -- it amounts to setting a default value)
>       * `PCMK_node_start_state`
>       * `PCMK_remote_port`
>       * `PCMK_respawned`

nrwahl2 commented on doc/sphinx/Pacemaker_Administration/agents.rst:
```diff
@@ -422,6 +422,100 @@ Exceptions to the recovery handling described above:
+     - Cluster-layer ID of the node where the current operation is running (or
+       node name for Pacemaker Remote nodes).
+   * - OCF_RESKEY_CRM_meta_physical_host
+     - If the node where the current operation is running is a bundle container,
```
actually if it's a guest node at all, as I noticed later

nrwahl2 commented on doc/sphinx/Pacemaker_Administration/options.rst:
```diff
@@ -0,0 +1,179 @@
+          single: environment variable; CIB_shadow_dir
+
+       CIB_shadow_dir
+     - (see description)
```
Sure. The reason is that I'd been saving the default column for literal values. But we're not always consistently doing that in other tables.

nrwahl2 commented on doc/sphinx/conf.py.in:
```diff
@@ -37,9 +37,10 @@ rst_prolog="""
 .. |PCMK_INIT_ENV_FILE| replace:: ``%PACEMAKER_CONFIG_DIR%/pcmk-init.env``
 .. |PCMK_LOG_FILE| replace:: %CRM_LOG_DIR%/pacemaker.log
-.. |PCMK_GNUTLS_PRIORITIES| replace:: %PCMK_GNUTLS_PRIORITIES%
+.. |PCMK_REMOTE_LOG_FILE| replace:: ``/var/log/pcmk-init.log``
```
Okay. It looks as if it's the default for all Pacemaker remote nodes if `PCMK__ENV_LOGFILE` is not set:
* `pacemaker-execd.c:main()` -> `remoted_spawn_pidone()` -> 
```
   if (pcmk__env_option(PCMK__ENV_LOGFILE) == NULL) {
        pcmk__set_env_option(PCMK__ENV_LOGFILE, "/var/log/pcmk-init.log", true);
    }
```

kgaillot replied:
Only if PID is 1 or PCMK__ENV_REMOTE_PID1 is "full", which effectively means only in a container (the developer testing case is not interesting to users)

nrwahl2 replied:
Ahh

nrwahl2 commented:
Updated as follows:
* Moved defaults for `CIB_shadow_dir` and `CIB_user` to defaults column
* Noted that `OCF_RESKEY_CRM_meta_physical_host` is set for all guest nodes, not just bundle containers
* Added references and index entries for OCF environment variables
* Noted that Pacemaker may set other env vars but that an agent should not rely on them
* Added missing `sed` expansion for `%CRM_CONFIG_DIR%`
* Modified note about `/var/log/pcmk-init.log` per https://github.com/ClusterLabs/pacemaker/pull/3277#discussion_r1439071231

kgaillot commented on doc/sphinx/Pacemaker_Explained/local-options.rst:
```diff
@@ -81,7 +93,10 @@ environment variables when Pacemaker daemons start up.
        severity. This log is of more use to developers and advanced system
-       administrators, and when reporting problems.
+       administrators, and when reporting problems. Note: The default is
+       |PCMK_REMOTE_LOG_FILE| (inside the container) bundled container nodes;
```
for

nrwahl2 replied:
thanks

kgaillot reviewed (commented):
looks good, let me know when you're ready to merge

nrwahl2 commented:
Ready.

kgaillot commented:
Whoops, resubmit against main branch
